fastapi/routing.py

```diff
@@ -921,6 +924,37 @@ def add_api_route(
             generate_unique_id_function=current_generate_unique_id,
         )
         self.routes.append(route)
+        if add_auto_options_route:
+            self._update_auto_options_routes(route, path)
+
+    def _update_auto_options_routes(self, new_route: APIRoute, path: str) -> None:
+        auto_options_index: Optional[int] = None
+        allowed_methods: Set[str] = set()
+        for index, route in enumerate(self.routes):
+            if isinstance(route, APIRoute):
+                if route.path == new_route.path:
+                    if hasattr(route, "is_auto_options") and route.is_auto_options:
+                        auto_options_index = index
+                    else:
+                        allowed_methods.update(route.methods)
+
+        if auto_options_index is not None:
+            self.routes.pop(auto_options_index)
+
+        if "OPTIONS" not in new_route.methods:
+
+            async def options_route() -> Response:
+                return Response(headers={"Allow": ", ".join(allowed_methods)})
+
+            self.routes.append(
+                APIRoute(
+                    self.prefix + path,
+                    endpoint=options_route,
+                    methods=["OPTIONS"],
+                    include_in_schema=True,
```

Comment:
Does this mean that it will be visible in the docs-UI? I.e. will each endpoint also have the corresponding endpoint with the options-method visible in the UI?